### ISSUE TITLE
Working with Homogeneous Records as Map

### ISSUE THREAD

**Issue by justinwoo** [dhall-lang/dhall-haskell#505]
I tried looking at this as a library, but I didn't find much to help me use homogeneous records as `Map Text a`.

A "solution" I found was to try to isolate the thing I wanted to read and handle its Expr directly, but I didn't find any nicer way to work with this as a member of a record.

**Comment by f-f** [comment]
[Relevant Twitter thread](https://twitter.com/jusrin00/status/1015542253803188224)

**Comment by sellout** [comment]
So, you’re looking for something like this?
```haskell
inputRecord :: Type a -> Text -> HashMap.InsOrd Text a
inputRecord ty =
  (\case
      RecordLit hm -> maybe (fail "wrong value type") pure (traverse (extract ty) hm)
      _ -> fail "Not a record")
  . inputExpr
```
(I haven’t tried that – just trying to sketch it out.)

**Comment by Gabriel439** [comment]
@justinwoo: I think it would be something like this:

```haskell
{-# LANGUAGE RecordWildCards #-}

import Data.Map (Map)
import Data.Text (Text)
import Dhall (Interpret(..))

import qualified Data.Map

data KeyValue a = KeyValue { mapKey :: Text, mapValue :: a }

toMap :: [KeyValue a] -> Map Text a
toMap keyValues = Data.Map.fromList (map adapt keyValues)
  where
    adapt (KeyValue {..}) = (mapKey, mapValue)

-- Wrap `Map` in a newtype if you want to avoid an orphan instance
instance Interpret a => Interpret (Map Text a) where
    autoWith options = fmap toMap (autoWith options)
```

**Comment by justinwoo** [comment]
@Gabriel439 is there a way to convert a homogeneous record to a list of keyvals? I use the record property when writing the actual Dhall code, but want to import it as a map when working with it

**Comment by f-f** [comment]
I believe we'd need the `toMap` function described in https://github.com/dhall-lang/dhall-lang/issues/169#issuecomment-395893113 and https://github.com/dhall-lang/dhall-lang/issues/164#issuecomment-398910371 to convert a Record to a list of keyvals

**Comment by Gabriel439** [comment]
Yeah, `toMap` is the change to the language I'd propose to support that

**Comment by blamario** [comment]
Let me just say that this is the last issue stopping me from converting my company's configuration files to Dhall. The workaround of using associative lists would make the configuration files both uglier and less type-safe.

Is there any unresolved design issue blocking `toMap` or is it just waiting for somebody to do the work?

**Comment by Gabriel439** [comment]
@blamario: Just waiting for somebody to do the work.  As far as I know there aren't any unresolved design issues

**Comment by blamario** [comment]
I may give it a try one of these days. I presume this would be new syntax, since there is no way to assign `toMap` a type?

**Comment by Gabriel439** [comment]
@blamario: Yeah, it would be new syntax

**Comment by blamario** [comment]
I gave this some thought, and I came up with a design issue after all. Instead of being a syntactic construct itself, `toMap` could be defined through another, more general syntactic construct - call it `reifyFields` for now:

    toMap = \recordType-> \record-> Prelude/map
               (\field-> {mapKey= field.mapKey,
                          mapValue= merge record field.mapValue})
               (reifyFields recordType)

The proposed `reifyFields` primitive would map a given record *type* to a list *value*. Examples:

    reifyFields {} = []
    reifyFields {a: Text} = [{mapKey= "a", mapValue= <a>.a}]
    reifyFields {a: Natural 1, b: Text} =
        [{mapKey= "a", mapValue= <a | b>.a},
         {mapKey= "b", mapValue= <a | b>.b}]

I can think of some uses for `reifyFields` that would go beyond `toMap`. However I'm not sure if it is implementable in Dhall, nor where Dhall's priorities lie between the generality and performance.

**Comment by blamario** [comment]
I realized immediately after the posting that `Prelude/map` requires the list item type as well. That would require another primitive `reifyFieldsType`:

    reifyFieldsType {} = {mapKey: Text, mapValue: <>}
    reifyFieldsType {a: Text} = {mapKey: Text, mapValue: <a>}
    reifyFieldsType {a: Natural 1, b: Text} = {mapKey: Text, mapValue: <a | b>}

**Comment by Gabriel439** [comment]
@blamario: If `toMap` were a first class function it's not clear to me what the type of `toMap` would be

Also, on the specific topic of reification, usually the reason we don't do reification in Dhall is because Dhall doesn't have a strong emphasis on meta-programming.  Instead, we usually do metaprogramming within the host language that binds to Dhall.

Another thing to keep in mind is that sometimes we redesign features when we discover better ways of doing things.  A good example of this is Dhall's support for union literals, which underwent 3 iterations before we settled on something we liked.  So sometimes it's fine to get something that's simple at first and if it needs to be generalized or improved later on we can deprecate the old way of doing things and provide automatic migrations to the new way of doing things as necessary.

**Comment by blamario** [comment]
> If `toMap` were a first class function it's not clear to me what the type of `toMap` would be

Looking back at my explanation, I really didn't make this clear: `toMap` wouldn't exist. Its definition I proposed above should be read as a guide to achieving the same effect. While `toMap` alone indeed couldn't be typed, no more than `reifyFields` could, an application of `reifyFields` to any record type would be well-typed.

> we usually do metaprogramming within the host language that binds to Dhall

For my use case, I would be perfectly happy to be able to somehow obtain the list of field names for a particular Dhall record type in Haskell, I have no need of it within Dhall. In other words, my only goal is to load a Dhall record as a Haskell `Map`. Adding `toMap` to Dhall was your preferred solution, if I understood correctly. But if meta-programming is a job for the host language, why not add the mechanism in the API instead?

**Comment by Gabriel439** [comment]
@blamario: I understand that `toMap` could be implemented in terms of other operators, but my understanding is that the primary use case for `toMap` is purely ergonomic.  In other words, the whole reason users want `toMap` is so that instead of writing this:

```dhall
[ { mapKey = "foo", mapValue = 1
  , mapKey = "bar", mapValue = 2
  }
]
```

... they can instead write:

```dhall
toMap { foo = 1, bar = 2 }
```

If they had to replace `toMap` with lower-level primitives like `reifyFields` then it no longer benefits that use case since they could just as well have written the association-list representation directly at the same syntactic cost.

Also, even if your goal were to programmatically retrieve record fields then I believe you would still want `toMap` to be the lower-level primitive anyway.  The reason why is that once you have the association list representation you can easily extract the keys and the values by using `Prelude/Map/keys` and `Prelude/Map/values`:

```dhall
let Prelude = https://prelude.dhall-lang.org/package.dhall

in  Prelude.Map.keys (toMap { foo = 1, bar = 2 })

-- = [ "foo", "bar" ]
```

**Comment by blamario** [comment]
> If they had to replace `toMap` with lower-level primitives like `reifyFields` then it no longer benefits that use case since they could just as well have written the association-list representation directly at the same syntactic cost.

Good point. It's not quite the *same* syntactic cost, because the association-list representation cost is linear rather than constant, but I agree with your objection.

You did not answer my other question though. I don't know about [dhall-lang/dhall-lang#169 (comment)](https://github.com/dhall-lang/dhall-lang/issues/169#issuecomment-395893113), but both the present issue and [dhall-lang/dhall-lang#164 (comment)](https://github.com/dhall-lang/dhall-lang/issues/164#issuecomment-398910371) could be solved outside Dhall itself.

In other words, can we simply loosen the API so that

    input auto "{a= 4, b= 1}" :: IO (Map Text Natural)

is valid in Haskell (and other host languages as applicable)? If you're concerned about injectivity, we could introduce a `newtype FieldMap a = FieldMap (Map Text a)` for the purpose. This would have the benefit of keeping Dhall cleaner, with fewer second-class reification constructs.

**Comment by Gabriel439** [comment]
@blamario: The main issue with addressing this within the `Interpret` instance for `Map` is that some configuration formats expect a list of weakly-typed records.  An example would be a JSON configuration file like this one (which is a common idiom in ops configurations):

```json
[ { "x": 1, "y": 2 }
, { "x": 1 }
]
```

If we model that in Dhall using `toMap`, we could write:

```dhall
[ toMap { x = 1, y = 2 }
, toMap { x = 1 }
] : List { mapKey : Text, mapValue : Natural }
```

... but we wouldn't be able to write:

```dhall
[ { x = 1, y = 2 }
, { x = 1 }
] : List ???
```

... because the expression does not type-check (due to the `List` element type-mismatch), and the Haskell binding won't attempt to marshal code that does not type-check.

**Comment by blamario** [comment]
I see. That's not something I intend to do, but I can understand the appeal. I guess that concludes my series of design questions. Thank you for the answers.

**Comment by blamario** [comment]
I think the implementation is pretty much done on my fork, but I'm running into some issues now that I set out to add tests. I guess I must be doing something wrong, but I can't find any developer documentation on the proper procedure. After adding the missing `<>` import to  `tests/Dhall/Test/Diff.hs` and `tests/Dhall/Test/Parser.hs` I get:

```
dhall-haskell/dhall$ cabal v2-test --enable-tests
Build profile: -w ghc-8.0.2 -O1
In order, the following will be built (use -v for more details):
 - dhall-1.24.0 (test:doctest) (ephemeral targets)
 - dhall-1.24.0 (test:tasty) (ephemeral targets)
Preprocessing test suite 'doctest' for dhall-1.24.0..
Building test suite 'doctest' for dhall-1.24.0..
Preprocessing test suite 'tasty' for dhall-1.24.0..
Building test suite 'tasty' for dhall-1.24.0..
Running 1 test suites...
Test suite doctest: RUNNING...

/home/mb/lab/gavia/dhall-haskell/dhall/src/Dhall/Parser.hs:47:0: error:
     error: missing binary operator before token "("
     #if MIN_VERSION_megaparsec(7, 0, 0)
     

/home/mb/lab/gavia/dhall-haskell/dhall/src/Dhall/Parser.hs:57:0: error:
     error: missing binary operator before token "("
     #if MIN_VERSION_megaparsec(7, 0, 0)
     
doctest: `gcc' failed in phase `C pre-processor'. (Exit code: 1)
Test suite doctest: FAIL
Test suite logged to:
/home/mb/lab/gavia/dhall-haskell/dist-newstyle/build/x86_64-linux/ghc-8.0.2/dhall-1.24.0/t/doctest/test/dhall-1.24.0-doctest.log
0 of 1 test suites (0 of 1 test cases) passed.
Running 1 test suites...
Test suite tasty: RUNNING...
tasty: ./dhall-lang/tests/normalization/success: fileAccess: does not exist (No such file or directory)
Test suite tasty: FAIL
Test suite logged to:
/home/mb/lab/gavia/dhall-haskell/dist-newstyle/build/x86_64-linux/ghc-8.0.2/dhall-1.24.0/t/tasty/test/dhall-1.24.0-tasty.log
0 of 1 test suites (0 of 1 test cases) passed.
cabal: Tests failed for test:doctest from dhall-1.24.0.
Tests failed for test:tasty from dhall-1.24.0.
```

**Comment by Gabriel439** [comment]
@blamario: Do you get the same test failure on a clean checkout of `master`?

As far as I can tell this looks like a `doctest` bug

**Comment by blamario** [comment]
Yes, `doctest` has the same issue on `master`. It's the latest version of `doctest` as well. It may be that it only works with a newer version of GHC, if newer GHC treats the pre-processor directives differently.

```
$ doctest --version
doctest version 0.16.1
using version 8.0.2 of the GHC API
using /usr/bin/ghc-8.0.2
```

I should also note that I'm running from within the `dhall-haskell/dhall` directory. When I try running `cabal new-build all` on a fresh clone of `master`, it ends with

```
src/Dhall/LSP/Util.hs:42:11: error:
    • The constructor ‘LSP.VirtualFile’ should have 3 arguments, but has been given 2
    • In the pattern: LSP.VirtualFile _ rope
      In the pattern: Just (LSP.VirtualFile _ rope)
      In a case alternative:
          Just (LSP.VirtualFile _ rope) -> return (Rope.toText rope)
```

Can you point me to any document that describes the test infrastructure? The `tasty` suite appears to be running tests from the `dhall-lang` repository, if I am not mistaken. But I can't be expected to add the language tests to that repository for a feature that's implemented by only one back-end: that would cause all the other implementations to fail the test suite. Or can I?

**Comment by Gabriel439** [comment]
@blamario: Actually, the expectation is that we will upstream this feature into the standard (and I can take care of the standardization process so that you don't have to).  We try to avoid implementation-specific language features.  So this would be added to the standard test suite in `dhall-lang`.

There are some features specific to the Haskell implementation outside of the language (like `dhall diff` and `dhall format`), which are supported by local test files, but tests for language functionality are upstreamed into `dhall-lang`.

Regarding the `lsp-server` issue, that looks like it is due to a missing bound on the `haskell-lsp` dependency

**Comment by blamario** [comment]
> So this would be added to the standard test suite in dhall-lang.

Ok, thanks for the clarification. That means I'll need to submit a PRs to each of the two repositories; should they go in any particular order?

> Regarding the lsp-server issue, that looks like it is due to a missing bound on the haskell-lsp dependency.

I can fix that.

**Comment by Gabriel439** [comment]
@blamario: Either order is fine.  Usually what I do is locally prototype a change to `dhall-haskell` to make sure that I didn't miss any details, then write up the standard proposal to ensure that the solution is constrained to be simple, then simultaneously put up both pull requests.

**Comment by Gabriel439** [comment]
Since we're getting close to standardizing this, I just wanted to double-check with everybody that people are okay with `toMap` as the name for this new keyword.  I personally don't have any objections (I was the one who originally suggested the name), but it's not too late to suggest another name if people have any concerns

**Comment by sjakobi** [comment]
What's the status of this issue now that `toMap` has been standardized (see https://github.com/dhall-lang/dhall-lang/releases/tag/v9.0.0)? Can it be closed?

**Comment by Gabriel439** [comment]
I'll close this since I'm reasonably certain that `toMap` addressed the original issue here.  Feel free to reopen if there's something else to do, though
